### ISSUE TITLE
hv_line~ - question about timing

### ISSUE THREAD

**Issue by jayrope** [enzienaudio/heavy#106]
I have  question about hv_line~: Is the abstraction provided here capable of sample exact timing? Usually, when i put in a value into vline in PD the default timing is in milliseconds. I wonder which default we have here, and how to achieve to work with a line or vline object either sample exact or at least down to block boundaries.

Thank you for any insight.

**Comment by diplojocus** [comment]
@jayrope are you talking about the `hv_vline~.pd` abstraction from here: https://github.com/enzienaudio/heavylib/blob/master/hv_vline%7E.pd

This internally uses the heavy SignalLine implementation found here: https://github.com/enzienaudio/heavy/blob/master/src/SignalLine.h#L70

One thing to bear in mind is that all messages in heavy are quantised to the minimum block size. This will be either 1, 4 or 8 depending on the optimisation flags set.

However, the calculation to the target output in the line object is sample accurate.

This means _when_ the message gets received might be determined by your min block size, but the time it takes get compute a particular ramp will be sample accurate.

**Comment by jayrope** [comment]
Okay, that teaches me something. 
I used [https://raw.githubusercontent.com/enzienaudio/heavylib/master/hv_vline~.pd](https://raw.githubusercontent.com/enzienaudio/heavylib/master/hv_vline~.pd)

• If i do [hv_vline~ 1] or feed a ramp with this timing value in into it - does this timing actually perform as one sample instead of one millisecond?

In terms of block size in PD i can only go down to something like 64, as otherwise audio interrupts are taking over. 

• So can i go down below that safely, knowing, that the OWL will still be able to process this glitch-free?

**Comment by diplojocus** [comment]
> If i do [hv_vline~ 1] or feed a ramp with this timing value in into it - does this timing actually perform as one sample instead of one millisecond?

No, it still takes ms, but you can calculate the sample timing by taking into account the sampling frequency.
`numSamplesInMs = numSamples * 1000 / sampleRate`

> So can i go down below that safely, knowing, that the OWL will still be able to process this glitch-free?

I can't remember off the top of my head if OWL supports NEON instructions but if it does it'll be running with an internal block size of 4 samples, otherwise it'll be 1. 

Obviously you'll be limited by how much processing you can do in that time, a more CPU intensive patch can result in dropouts.

One thing to note, is that these minimum block sizes are internal to heavy. Say if OWL is processing with a block size of 64, internally heavy would process 16 blocks of 4 (mainly because it's faster). What this will also mean is that messages sent to your patch will be quantised to this parent block size (64). But all internal messages in the patch can be assumed to be quantised to the minimum block amount (4).

Hope that helps!

**Comment by jayrope** [comment]
`No, it still takes ms, but you can calculate the sample timing by taking into account the sampling frequency.
numSamplesInMs = numSamples * 1000 / sampleRate`

So i can feed fractions of 1 instead to go down to sample level, like [1/44.1] or so?

**Comment by diplojocus** [comment]
Sure

**Comment by jayrope** [comment]
Perfect, and thank you a lot for the insight, Joe.

**Comment by HerrGustav** [comment]
Is the here mentioned "hv_line~"  a workaround for the [vline~] object? 

I'm trying to get a simple ADSR - envelope generator running and my first solution with vline worked pretty good. 
But since heavy doesn't support vline I tried to model something in the same direction using the [line~] - object. 
Sadly there are audible clicks right now when I change the frequency of the connected synth/oscillator.

What is the best workaround in heavy for vline? Sorry I'm quite new to PD as well.

**Comment by mhroth** [comment]
@HerrGustav You can use the `hv.vline` object available in the heavylib repository as a suitable replacement. https://github.com/enzienaudio/heavylib

**Comment by HerrGustav** [comment]
@mhroth Thanks a lot!  That worked pretty flawless. 
I didn't find these objects mentioned in the documentation.
